craft_parts/executor/part_handler.py

```diff
@@ -0,0 +1,384 @@
+
+from craft_parts import callbacks, errors, packages, plugins, sources
+from craft_parts.actions import Action, ActionType
+from craft_parts.infos import PartInfo, StepInfo
+from craft_parts.packages import errors as packages_errors
+from craft_parts.parts import Part
+from craft_parts.plugins import Plugin
+from craft_parts.state_manager import states
+from craft_parts.state_manager.states import StepState
+from craft_parts.steps import Step
+from craft_parts.utils import os_utils
+
+from .organize import organize_files
+from .step_handler import StepContents, StepHandler
+
+logger = logging.getLogger(__name__)
+
+
+class PartHandler:
+    """Handle steps for a part using the appropriate plugins.
+
+    :param part: The part being processed.
+    :param part_info: Information about the part being processed.
+    :param part_list: A list containing all parts.
+    """
+
+    def __init__(
+        self,
+        part: Part,
+        *,
+        part_info: PartInfo,
+        part_list: List[Part],
+    ):
+        self._part = part
+        self._part_info = part_info
+        self._part_list = part_list
+
+        self._plugin = plugins.get_plugin(
+            part=part,
+            properties=part.plugin_properties,
+            part_info=part_info,
+        )
+
+        self._part_properties = part.spec.marshal()
+        self._source_handler = sources.get_source_handler(
+            application_name=part_info.application_name,
+            part=part,
+            project_dirs=part_info.dirs,
+        )
+
+        self._build_packages = _get_build_packages(part=self._part, plugin=self._plugin)
+        self._build_snaps = _get_build_snaps(part=self._part, plugin=self._plugin)
+
+    def run_action(self, action: Action) -> None:
+        """Execute the given action for this part using a plugin.
+
+        :param action: The action to execute.
+        """
+        step_info = StepInfo(self._part_info, action.step)
+
+        if action.action_type == ActionType.UPDATE:
+            # TODO: add update action handler
+            return
+
+        if action.action_type == ActionType.RERUN:
+            # TODO: clean step
+            pass
+
+        handler: Callable[[StepInfo], StepState]
+
+        if action.step == Step.PULL:
+            handler = self._run_pull
+        elif action.step == Step.BUILD:
+            handler = self._run_build
+        elif action.step == Step.STAGE:
+            handler = self._run_stage
+        elif action.step == Step.PRIME:
+            handler = self._run_prime
+        else:
+            raise RuntimeError("cannot run action for invalid step {action.step!r}")
+
+        callbacks.run_pre_step(step_info)
+        state = handler(step_info)
+        state_file = states.state_file_path(self._part, action.step)
+        state.write(state_file)
+        callbacks.run_post_step(step_info)
+
+    def _run_pull(self, step_info: StepInfo) -> StepState:
+        _remove(self._part.part_src_dir)
```

Comment:
are we ever going to consider a reentrant "pull" that updates sources? (life git pull multiple times?)

Reply:
This could save time and bandwidth if e.g. git sources are being changed locally, but would also require manual cleaning when starting over. Currently we support updating for certain sources (local. git) which should cover the "locally changed sources" scenario. Update support is scheduled to be added in the forthcoming Part Handler (2/2) PR.